### ISSUE TITLE
Unable to restore a layer of class TextVectorization - Text Classification 

### ISSUE THREAD

**Issue by alihasn92** [tensorflow/tensorflow#45231]
**System information**
Google Colab

**When I run the example provided by official tensorflow Basic text classification, everything runs fine until model  save. But when I load the model it gives me this error. 

RuntimeError: Unable to restore a layer of class TextVectorization. Layers of class TextVectorization require that the class be provided to the model loading code, either by registering the class using @keras.utils.register_keras_serializable on the class def and including that file in your program, or by passing the class in a keras.utils.CustomObjectScope that wraps this load call.
**

**Model should be loaded successfully and process raw input**

**https://colab.research.google.com/gist/amahendrakar/8b65a688dc87ce9ca07ffb0ce50b84c7/44199.ipynb#scrollTo=fEjmSrKIqiiM**

**Example Link: https://tensorflow.google.cn/tutorials/keras/text_classification** 

There is another issue [(#44199)](https://github.com/tensorflow/tensorflow/issues/44199) like this but it was closed abruptly without any solution so that's why I have to create a new issue as I did not receive any response. Thank you

**Comment by Saduf2019** [comment]
I am able to replicate the issue reported, please find the [gist here](https://colab.research.google.com/gist/Saduf2019/24dd76c7a8f04e54dd5bd8ad6e937c6d/untitled472.ipynb).

**Comment by ymodak** [comment]
I was able to execute the script successfully using the hack mentioned in tagged github issue above. Find the [gist](https://colab.research.google.com/gist/ymodak/afb4b5015dd5b2abf017d1a79da6ee60/44199.ipynb) here.
```python
@tf.keras.utils.register_keras_serializable()
def custom_standardization(input_data):
 ...
```

**Comment by alihasn92** [comment]
> I was able to execute the script successfully using the hack mentioned in tagged github issue above. Find the [gist](https://colab.research.google.com/gist/ymodak/afb4b5015dd5b2abf017d1a79da6ee60/44199.ipynb) here.
> 
> ```python
> @tf.keras.utils.register_keras_serializable()
> def custom_standardization(input_data):
>  ...
> ```

Thank you!
Let me try it again, if all good, I will close the issue.

**Comment by alihasn92** [comment]
I just tested it, it does work only if you save the model and load back in the same notebook like you did. As soon as you download the whole model folder and try to load it in new notebook, it starts giving same error.

![image](https://user-images.githubusercontent.com/23615976/100745057-020c0d80-33df-11eb-9e1c-c289a5775342.png)

Find [gist](https://colab.research.google.com/drive/1zNt8yz_mdW9JhZ0cHULr_lgYb_Qj9g_8?usp=sharing) here.

**Comment by alihasn92** [comment]
Do we have any update here?

**Comment by alihasn92** [comment]
Hello, Everyone!
There is no response on this issue in two weeks. Could someone kindly update the status here? Thank you

**Comment by alihasn92** [comment]
Push +1

**Comment by linchpinlin** [comment]
l have the same problem./(ㄒoㄒ)/~~

**Comment by linchpinlin** [comment]
In your new notebook, you can try the following:
"""
import re
import string

@tf.keras.utils.register_keras_serializable()
def custom_standardization(input_data):
  lowercase = tf.strings.lower(input_data)
  stripped_html = tf.strings.regex_replace(lowercase, '<br />', ' ')
  return tf.strings.regex_replace(stripped_html,
                                  '[%s]' % re.escape(string.punctuation), '')

"""
Then l reload the model successfully.

**Comment by alihasn92** [comment]
But why do we need to preprocess the incoming text again? Input for the prediction should be raw!

**Comment by CameronWhittleCT** [comment]
I have run into the same issue following this example, I agree that the work around is not a true solution. Documentation examples should run without errors. 

Even with the workaround I received this warning when having the model make predictions on raw text:
WARNING:tensorflow:5 out of the last 10 calls to <function Model.make_predict_function.<locals>.predict_function at 0x00000285CEFBCEE8> triggered tf.function retracing. Tracing is expensive and the excessive number of tracings could be due to (1) creating @tf.function repeatedly in a loop, (2) passing tensors with different shapes, (3) passing Python objects instead of tensors. For (1), please define your @tf.function outside of the loop. For (2), @tf.function has experimental_relax_shapes=True option that relaxes argument shapes that can avoid unnecessary retracing. For (3), please refer to https://www.tensorflow.org/guide/function#controlling_retracing and https://www.tensorflow.org/api_docs/python/tf/function for  more details.

Which I suspect is due to inappropriate use of 
@tf.keras.utils.register_keras_serializable()
in the workaround

**Comment by alihasn92** [comment]
> I have run into the same issue following this example, I agree that the work around is not a true solution. Documentation examples should run without errors.
> 
> Even with the workaround I received this warning when having the model make predictions on raw text:
> WARNING:tensorflow:5 out of the last 10 calls to <function Model.make_predict_function..predict_function at 0x00000285CEFBCEE8> triggered tf.function retracing. Tracing is expensive and the excessive number of tracings could be due to (1) creating @tf.function repeatedly in a loop, (2) passing tensors with different shapes, (3) passing Python objects instead of tensors. For (1), please define your @tf.function outside of the loop. For (2), @tf.function has experimental_relax_shapes=True option that relaxes argument shapes that can avoid unnecessary retracing. For (3), please refer to https://www.tensorflow.org/guide/function#controlling_retracing and https://www.tensorflow.org/api_docs/python/tf/function for more details.
> 
> Which I suspect is due to inappropriate use of
> @tf.keras.utils.register_keras_serializable()
> in the workaround

Could you link the notebook where you were able to run the example and load model etc? Because for me, those example did not work at all (for loading part) when you try to load model into new notebook. Thank you

**Comment by CameronWhittleCT** [comment]
In truth my code is quite different from the example, but I did use it as guidance for writing a custom standardization function as part of a TextVectorization layer. 
I assure you that my model saved and loaded perfectly before changing from lower_and_strip_punctuation to my own custom standardization in the TextVectorization layer. After switching I had the same RuntimeError you encountered in the head of this page.

**Comment by DanielMH** [comment]
I also ran into this error message when I implemented (and customized) the code from the "Basic Text Classification" tutorial.

Instead of running the code in a notebook, I have two scripts, one for building, training and saving the model and the other one for loading it and making predictions.

Implementing the comments above solved the issue for me. Because it was confusing to me, just to sum up what I had to do in order to get the script with the predictions running:

1. I added this first line in the first script before the function definition and saved the model again

```
@tf.keras.utils.register_keras_serializable()
def custom_standardization(input_data):
[...]
```

2. I also had to add the same line and the whole function definition in the second script to make sure that it works when I restart(!) ipython (where I currently run the scripts).

**Comment by DanielMH** [comment]
Alternatively, you can just use the default standardization method in the vectorizer layer when building the model:

```
vectorize_layer = TextVectorization(
        standardize='lower_and_strip_punctuation',
        max_tokens=max_features,
        output_mode='int',
        output_sequence_length=sequence_length)
```

**Comment by ewotawa** [comment]
Thanks for the information on this thread. I'm trying to do something similar to what @alihasn92 described. I'm building a project off of the documentation from the same TensorFlow tutorial. 

You can see how I train, test, and save the model in the following notebook:
https://github.com/ewotawa/top_100_wines_2020_ML/blob/master/WS%20Top%20100%202020%20-%20Classifier.ipynb

To continue with my project, I need to import the model in a different location and let it evaluate new raw text. I tested what I thought was the correct approach in the following notebook:
https://github.com/ewotawa/top_100_wines_2020_ML/blob/master/WS%20Top%20100%202020%20-%20Classifier%20-%20Test%20Load.ipynb

Using the workaround, I can load the model and print the model summary, but when I use predict on new raw sample text, the error returned is below:
`IndexError: list index out of range` 

I'd be grateful for your thoughts on what steps I'm missing. Thanks much.

**Comment by sachinprasadhs** [comment]
@statneoproteek , you can solve the issue by following the steps as suggested in the error. `RuntimeError: Unable to restore a layer of class TextVectorization. Layers of class TextVectorization require that the class be provided to the model loading code, either by registering the class using @keras.utils.register_keras_serializable on the class def and including that file in your program, or by passing the class in a keras.utils.CustomObjectScope that wraps this load call.`

Attaching the working [code](https://colab.research.google.com/gist/sachinprasadhs/9471d05206aa1c206584c33a9d5d598f/44199.ipynb) with the above changes as reference. Thanks!

**Comment by statneoproteek** [comment]
Deserialization of the TextVectorization layer shouldn't  be dependent on the class to be present in the model loading code.  It definitely does not facilitate sharing models in a clean way. 
Please correct me if I am wrong: Wouldn't passing the class in a keras.utils.CustomObjectScope would also need the class to be present in the model loading code?

TL;DR: I think since the layer does not serialize completely, it warrants a fix.
I am not sure if this is the github issue to handle this. Please feel free to point it out if that's the case.

**Comment by sachinprasadhs** [comment]
@statneoproteek, This issue has been moved to keras-team/keras repo, you can track the further details using above linked issue request. Thanks!

**Comment by luficerg** [comment]
Soory to say but it is not so solved in the tensorflow basics itself, plz have a look https://colab.research.google.com/github/tensorflow/docs/blob/master/site/en/guide/intro_to_modules.ipynb#scrollTo=Wj5DW-LCopry

and I think it is some internal prob

**Comment by amrc66283** [comment]
I was having the same issue, although the above solution is working but what matters is when we are trying to save the model from one place and loading it on other place make sure to use the same TensorFlow version, I did it for me and it started to work.

shortly, make sure to use same version of tensorflow. @alihasn92

**Comment by Venon282** [comment]
Problem was open in 2020 and still no solution ?
